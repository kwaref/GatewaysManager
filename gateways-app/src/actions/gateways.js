import Swal from 'sweetalert2'
import { fetchWithOutToken } from '../helpers/fetch'
import { prepareGateways } from '../helpers/prepareGateways'
import { types } from '../types/types'
import { uiStopLoading } from './ui'

export const gatewayStartAddNew = gateway => {
	return async dispatch => {
		try {
			const resp = await fetchWithOutToken('gateways', gateway, 'POST')
			const body = await resp.json()

			if (body.ok) {
				dispatch(gatewayAddNew(gateway))
				dispatch(gatewayStartLoading())
				dispatch(uiStopLoading())
			}
		} catch (error) {
			dispatch(uiStopLoading())
			console.log(error)
			Swal.fire('', error, 'error')
		}
	}
}

export const gatewayAddNew = gateway => ({
	type: types.gatewayAddNew,
	payload: gateway,
})
export const gatewaySetActive = gateway => ({
	type: types.gatewaySetActive,
	payload: gateway,
})
export const gatewayClearActive = () => ({
	type: types.gatewayClearActive,
})

export const gatewayStartUpdate = gateway => {
	return async dispatch => {
		try {
			const resp = await fetchWithOutToken(
				`gateways/${gateway._id}`,
				gateway,
				'PUT',
			)

			const body = await resp.json()
			if (body.ok) {
				dispatch(gatewayUpdated(gateway))
				dispatch(uiStopLoading())
			} else {
				dispatch(uiStopLoading())
				Swal.fire('Error', body.msg, 'error')
			}
		} catch (error) {
			dispatch(uiStopLoading())
			console.log(error)
			Swal.fire('', error, 'error')
		}
	}
}

export const gatewayUpdated = gateway => ({
	type: types.gatewayUpdated,
	payload: gateway,
})
export const gatewayStartLoading = () => {
	return async dispatch => {
		try {
			const resp = await fetchWithOutToken('gateways')
			const body = await resp.json()

			const gateways = prepareGateways(body.gateways)

			dispatch(gatewayLoaded(gateways))
			dispatch(uiStopLoading())
		} catch (error) {
			dispatch(uiStopLoading())
			console.log(error)
			Swal.fire('', error, 'error')
		}
	}
}

export const gatewayLoaded = gateways => ({
	type: types.gatewayLoaded,
	payload: gateways,
})

export const gatewayDeleted = gateway => ({
	type: types.gatewayDeleted,
	payload: gateway,
})

export const gatewayStartDelete = gateway => {
	return async dispatch => {
		try {
			const resp = await fetchWithOutToken(
				`gateways/${gateway._id}`,
				{},
				'DELETE',
			)

			const body = await resp.json()
			if (body.ok) {
				dispatch(gatewayDeleted(gateway))
				dispatch(uiStopLoading())
			} else {
				dispatch(uiStopLoading())
				Swal.fire('Error', body.msg, 'error')
			}
		} catch (error) {
			dispatch(uiStopLoading())
			console.log(error)
			Swal.fire('', error, 'error')
		}
	}
}
