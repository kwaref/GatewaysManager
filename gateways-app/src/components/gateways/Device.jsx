import React from 'react'
import { useDispatch } from 'react-redux'
import { Navigate, useParams } from 'react-router-dom'
import { gatewayStartUpdate } from '../../actions/gateways'
import { uiStartLoading } from '../../actions/ui'
import { useGetGatewayById } from '../../selectors/useGetGatewayById'

export const Device = ({ UID, vendor, date, status }) => {
	const dispatch = useDispatch()
	const { gatewayId } = useParams()

	const gateway = useGetGatewayById(gatewayId)

	if (!gateway) {
		return <Navigate to='/' />
	}

	const { devices } = gateway

	const onDelete = () => {
		dispatch(
			gatewayStartUpdate({
				...gateway,
				devices: devices.filter(device => device.UID !== UID),
			}),
		)
		dispatch(uiStartLoading())
	}

	return (
		<div className='card ms-1' style={{ maxWidth: 240, height: 240 }}>
			<div className='row no-gutters'>
				<div className='card-body'>
					<h3 className='card-title'>{UID}</h3>
					<hr />
					<p className='card-text'>Vendor: {vendor}</p>
					<p className='card-text'>Date: {date}</p>
					<p className='card-text'>Status: {status ? 'online' : 'offline'}</p>
					<i
						className='fas fa-trash-alt float-right'
						style={{ cursor: 'pointer' }}
						onClick={onDelete}></i>
				</div>
			</div>
		</div>
	)
}
