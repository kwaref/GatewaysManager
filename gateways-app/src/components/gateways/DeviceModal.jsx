import React, { useEffect, useState } from 'react'
import DatePicker from 'react-datepicker'
import Modal from 'react-modal'
import moment from 'moment'
import { useDispatch } from 'react-redux'
import { useSelector } from 'react-redux'
import { useParams } from 'react-router-dom'
import { gatewayStartUpdate } from '../../actions/gateways'
import { uiCloseDeviceModal, uiStartLoading } from '../../actions/ui'
import { useGetGatewayById } from '../../selectors/useGetGatewayById'
//import { ToggleSwitch } from '../ui/ToggleSwitch'
import 'react-datepicker/dist/react-datepicker.css'
import Swal from 'sweetalert2'
import { uidIsInDB } from '../../helpers/searchers'

const customStyles = {
	content: {
		top: '50%',
		left: '50%',
		right: 'auto',
		bottom: 'auto',
		marginRight: '-50%',
		transform: 'translate(-50%, -50%)',
	},
}
if (process.env.NODE_ENV !== 'test') Modal.setAppElement('#root')

const initDevice = {
	UID: '',
	vendor: '',
	date: moment().toDate(),
	status: true,
}

export const DeviceModal = () => {
	const { modalDeviceOpen } = useSelector(state => state.ui)
	const { gateways } = useSelector(state => state.gateway)

	const { gatewayId } = useParams()
	const gateway = useGetGatewayById(gatewayId)
	const { devices } = gateway

	const dispatch = useDispatch()

	const [formValues, setFormValues] = useState(initDevice)
	const { UID, vendor, status } = formValues
	const [startDate, setStartDate] = useState(moment().toDate())

	const [validUID, setValidUID] = useState(0)
	const [validVendor, setValidVendor] = useState(0)
	const [validDate, setValidDate] = useState(0)

	const handleSelectChange = ({ target }) => {
		const st = target.value === 'true' ? true : false

		setFormValues({
			...formValues,
			status: st,
		})
	}

	const handleInputChange = ({ target }) => {
		setFormValues({
			...formValues,
			[target.name]: target.value,
		})
	}

	const onDateChange = e => {
		setStartDate(e)

		setFormValues({
			...formValues,
			date: moment(e).format('MM/DD/YYYY'),
		})
	}

	useEffect(() => {
		if (!uidIsInDB(UID, gateways) && UID.trim().length > 0) {
			setValidUID(1)
		}
		if (UID.trim().length === 0) {
			setValidUID(0)
		}
		if (uidIsInDB(UID, gateways)) {
			setValidUID(-1)
		}
		if (vendor.trim().length > 0) {
			setValidVendor(1)
		} else {
			setValidVendor(0)
		}
		if (!startDate) {
			setValidDate(-1)
		} else {
			setValidDate(0)
		}
	}, [UID, vendor, startDate, gateways])

	const closeModal = () => {
		dispatch(uiCloseDeviceModal())
		setFormValues(initDevice)
	}

	const handleSubmit = e => {
		e.preventDefault()

		if (UID.trim().length < 1) {
			Swal.fire('', 'UID is required!', 'error')
			return setValidUID(-1)
		}
		if (uidIsInDB(UID, gateways)) {
			Swal.fire('', 'That UID is already in the DB!', 'error')
			return setValidUID(-1)
		}
		if (vendor.trim().length < 1) {
			Swal.fire('', "You must enter a Vendor's name", 'error')
			return setValidVendor(-1)
		}
		if (!startDate) {
			Swal.fire('', 'You must select a correct Date', 'error')
			return setValidDate(-1)
		}

		dispatch(
			gatewayStartUpdate({
				...gateway,
				devices: [
					...devices,
					{ ...formValues, date: moment(startDate).format('MM/DD/YYYY') },
				],
			}),
		)

		dispatch(uiStartLoading())

		setValidUID(0)
		setValidVendor(0)
		setValidDate(0)

		closeModal()
	}

	return (
		<Modal
			isOpen={modalDeviceOpen}
			onRequestClose={closeModal}
			style={customStyles}
			closeTimeoutMS={200}>
			<h1> New Device </h1>
			<hr />
			<form className='container' onSubmit={handleSubmit}>
				<div className='form-group'>
					<label>UID</label>
					<input
						type='text'
						className={`form-control ${validUID === 1 && 'is-valid'} ${
							validUID === -1 && 'is-invalid'
						}`}
						placeholder='UID'
						name='UID'
						autoComplete='off'
						value={UID}
						onChange={handleInputChange}
					/>
				</div>

				<div className='form-group'>
					<label>Vendor</label>
					<input
						className={`form-control ${validVendor === 1 && 'is-valid'} ${
							validVendor === -1 && 'is-invalid'
						}`}
						placeholder='Vendor'
						name='vendor'
						autoComplete='off'
						value={vendor}
						onChange={handleInputChange}
					/>
				</div>
				<div className='form-group'>
					<label>Date</label>
					<br />
					<DatePicker
						selected={startDate}
						onChange={date => onDateChange(date)}
						value={startDate}
						className={`form-control ${validDate === 1 && 'is-valid'} ${
							validDate === -1 && 'is-invalid'
						}`}
					/>
				</div>
				<div className='form-group'>
					<label>Status</label>
					<select
						className='form-control'
						name='status'
						value={status}
						onChange={handleSelectChange}>
						<option value={true}>online</option>
						<option value={false}>offline</option>
					</select>
				</div>

				<hr />

				<button type='submit' className='btn btn-outline-primary btn-block'>
					<i className='far fa-save'></i>
					<span> Save</span>
				</button>
			</form>
		</Modal>
	)
}
