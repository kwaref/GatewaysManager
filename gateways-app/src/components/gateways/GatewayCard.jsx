import React from 'react'
import { useDispatch } from 'react-redux'
import { Link } from 'react-router-dom'
import { gatewayStartDelete } from '../../actions/gateways'
import { uiStartLoading } from '../../actions/ui'

export const GatewayCard = ({ gateway }) => {
	//const { gateways } = useSelector(state => state.gateway)
	const dispatch = useDispatch()

	const onDelete = e => {
		dispatch(gatewayStartDelete(gateway))
		dispatch(uiStartLoading())
	}

	return (
		<div className='card ms-3' style={{ maxWidth: 240, height: 220 }}>
			<div className='row no-gutters'>
				<div className='card-body'>
					<h3 className='card-title truncate'>{gateway.name}</h3>
					<hr />
					<p className='card-text'>Serial: {gateway.serial}</p>
					<p className='card-text'>IP Address: {gateway.ipAddress}</p>
					<Link to={`./gateway/${gateway.serial}`}>
						<small>
							<i>More...</i>
						</small>
					</Link>
					<i
						className='fas fa-trash-alt float-right'
						style={{ cursor: 'pointer' }}
						onClick={onDelete}></i>
				</div>
			</div>
		</div>
	)
}
