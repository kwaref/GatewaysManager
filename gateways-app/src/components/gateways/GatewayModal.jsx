import React, { useEffect, useState } from 'react'
import Modal from 'react-modal'
import { useDispatch } from 'react-redux'
import { useSelector } from 'react-redux'
import Swal from 'sweetalert2'
import { gatewayStartAddNew } from '../../actions/gateways'
import { uiCloseGatewayModal, uiStartLoading } from '../../actions/ui'
import { isIP } from '../../helpers/isIP'
import { serialIsInDB } from '../../helpers/searchers'

const customStyles = {
	content: {
		top: '50%',
		left: '50%',
		right: 'auto',
		bottom: 'auto',
		marginRight: '-50%',
		transform: 'translate(-50%, -50%)',
	},
}
if (process.env.NODE_ENV !== 'test') Modal.setAppElement('#root')

const initGateway = {
	name: '',
	serial: '',
	ipAddress: '',
}

export const GatewayModal = () => {
	const { modalGatewayOpen } = useSelector(state => state.ui)
	const { gateways } = useSelector(state => state.gateway)
	const dispatch = useDispatch()
	const [validName, setValidName] = useState(0)
	const [validSerial, setValidSerial] = useState(0)
	const [validIP, setValidIP] = useState(0)
	const [formValues, setFormValues] = useState(initGateway)
	const { name, serial, ipAddress } = formValues

	const handleInputChange = ({ target }) => {
		setFormValues({
			...formValues,
			[target.name]: target.value,
		})
	}

	useEffect(() => {
		if (isIP(ipAddress)) {
			setValidIP(1)
		}
		if (!serialIsInDB(serial, gateways) && serial.trim().length > 0) {
			setValidSerial(1)
		}
		if (serial.trim().length === 0) {
			setValidSerial(0)
		}
		if (serialIsInDB(serial, gateways)) {
			setValidSerial(-1)
		}
		if (name.trim().length > 0) {
			setValidName(1)
		} else {
			setValidName(0)
		}
	}, [name, ipAddress, serial, gateways])

	const closeModal = () => {
		dispatch(uiCloseGatewayModal())
		setFormValues(initGateway)
	}

	const handleSubmit = e => {
		e.preventDefault()

		if (serial.trim().length < 1) {
			Swal.fire('', 'The serial is required', 'error')
			return setValidSerial(-1)
		}
		if (serialIsInDB(serial, gateways)) {
			Swal.fire('', 'That serial already exists in DB', 'error')
			return setValidSerial(-1)
		}
		if (name.trim().length < 2) {
			Swal.fire('', 'You must enter a name.', 'error')
			return setValidName(-1)
		}
		if (!isIP(ipAddress)) {
			Swal.fire('', 'You must enter a valid IP address', 'error')
			return setValidIP(-1)
		}

		dispatch(
			gatewayStartAddNew({
				...formValues,
				devices: [],
			}),
		)
		dispatch(uiStartLoading())

		setValidSerial(0)
		setValidName(0)
		setValidIP(0)

		closeModal()
	}

	return (
		<Modal
			isOpen={modalGatewayOpen}
			// onAfterOpen={afterOpenModal}
			onRequestClose={closeModal}
			style={customStyles}
			closeTimeoutMS={200}>
			<h1> New Gateway </h1>
			<hr />
			<form className='container' onSubmit={handleSubmit}>
				<div className='form-group'>
					<label>Serial number</label>
					<input
						type='text'
						className={`form-control ${validSerial === 1 && 'is-valid'} ${
							validSerial === -1 && 'is-invalid'
						}`}
						placeholder='Serial number'
						name='serial'
						autoComplete='off'
						value={serial}
						onChange={handleInputChange}
					/>
				</div>

				<div className='form-group'>
					<label>Name</label>
					<input
						className={`form-control ${validName === 1 && 'is-valid'} ${
							validName === -1 && 'is-invalid'
						}`}
						placeholder='Name'
						name='name'
						autoComplete='off'
						value={name}
						onChange={handleInputChange}
					/>
				</div>
				<div className='form-group'>
					<label>IP Address</label>
					<input
						className={`form-control ${validIP === 1 && 'is-valid'} ${
							validIP === -1 && 'is-invalid'
						}`}
						placeholder='IP Address'
						name='ipAddress'
						autoComplete='off'
						value={ipAddress}
						onChange={handleInputChange}
					/>
				</div>

				<hr />

				<button type='submit' className='btn btn-outline-primary btn-block'>
					<i className='far fa-save'></i>
					<span> Save</span>
				</button>
			</form>
		</Modal>
	)
}
