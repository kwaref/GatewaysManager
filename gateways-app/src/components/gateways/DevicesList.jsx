import React from 'react'
import { Device } from './Device'

export const DevicesList = ({ devices }) => {
	return (
		<div className='card-columns'>
			{devices.map(device => (
				<Device key={device.UID} {...device} />
			))}
		</div>
	)
}
