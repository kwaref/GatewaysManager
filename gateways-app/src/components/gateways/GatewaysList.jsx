import React from 'react'
import { useDispatch } from 'react-redux'
import { uiOpenGatewayModal } from '../../actions/ui'
import { GatewayCard } from './GatewayCard'

export const GatewaysList = ({ gateways }) => {
	const dispatch = useDispatch()

	// useEffect(() => {
	// 	dispatch(gatewayClearActive())
	// }, [dispatch])

	const handleOnClick = e => {
		dispatch(uiOpenGatewayModal())
	}

	return (
		<>
			<div className='card-columns'>
				{gateways.map(gateway => (
					<GatewayCard key={gateway.serial} gateway={gateway} />
				))}
			</div>
			<i
				className='fas fa-plus-circle fa-3x'
				style={{ cursor: 'pointer' }}
				onClick={handleOnClick}></i>
		</>
	)
}
