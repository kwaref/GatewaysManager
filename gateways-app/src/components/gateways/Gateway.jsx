import React, { useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { Navigate, useParams } from 'react-router-dom'
import { gatewaySetActive } from '../../actions/gateways'
import { uiOpenDeviceModal } from '../../actions/ui'
import { useGetGatewayById } from '../../selectors/useGetGatewayById'
import { DeviceModal } from './DeviceModal'
import { DevicesList } from './DevicesList'

export const Gateway = () => {
	const dispatch = useDispatch()
	const { gatewayId } = useParams()
	const handleClick = e => {
		dispatch(uiOpenDeviceModal())
	}
	const gateway = useGetGatewayById(gatewayId)
	useEffect(() => {
		dispatch(gatewaySetActive(gateway))
	}, [dispatch, gateway])
	if (!gateway) {
		return <Navigate to='/' />
	}

	const { devices } = gateway

	return (
		<div>
			<h1>{gateway.name}</h1>
			<hr />
			<h3>Serial: {gateway.serial}</h3>
			<h3>IP Address: {gateway.ipAddress}</h3>
			<hr />
			<h3>Devices:</h3>
			{devices && <DevicesList devices={gateway.devices} />}

			{devices.length < 10 && (
				<i
					className='fas fa-plus-circle fa-3x'
					style={{ cursor: 'pointer' }}
					onClick={handleClick}></i>
			)}

			<DeviceModal />
		</div>
	)
}
