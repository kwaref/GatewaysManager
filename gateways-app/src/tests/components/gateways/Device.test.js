import React from 'react'
import { shallow } from 'enzyme'
import { Device } from '../../../components/gateways/Device'
import { Provider } from 'react-redux'

import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import '@testing-library/jest-dom'

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {}
const store = mockStore(initState)
store.dispatch = jest.fn()

const props = { UID: '123', vendor: 'hp', status: true }

const wrapper = shallow(
	<Provider store={store}>
		<Device {...props} />
	</Provider>,
)

describe('Tests on Device component', () => {
	it('should be correctly rendered', () => {
		expect(wrapper).toMatchSnapshot()
	})
})
