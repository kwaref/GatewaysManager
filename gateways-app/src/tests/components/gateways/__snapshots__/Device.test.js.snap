// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on Device component should be correctly rendered 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "onStateChange": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <Device
    UID="123"
    status={true}
    vendor="hp"
  />
</ContextProvider>
`;
